### PR TITLE
Add option to crack using msbuild project graph

### PR DESCRIPTION
**Pull request by TheAngryByrd**
This PR creates a WorkspaceLoader based on [MSBuild's ProjectGraph](https://github.com/dotnet/msbuild/blob/40beb5a025fb18c82a159f407b5ce2383531a8eb/src/Build/Graph/ProjectGraph.cs).  This allows for MSBuild to do all the lookups/builds efficiently as it can for a given workspace's project graph rather than relying on building individually and caching them.  In my work project, initial load times went from 12 seconds to 7 seconds.

I've created a `IWorkspaceLoader` interface which will allow swapping of this so we can test in the field to find any weird edge cases this hasn't take into account.  See PRs the FSAC and Ionide for the feature flag changes. 

---

Currently I still have _some_ things left to finish.  

- [x] Using customProperties
- [x] generateBinLog
- [x] Equivalent tests for `WorkspaceLoaderViaProjectGraph` and fix current ones
- [x] Cleanup/refactor

### CODE REVIEW

File: build.fsx [hunk 1]
```diff
@@ -90,6 +90,18 @@ let DoNothing = ignore
 
 Target.create "Clean" (fun _ -> Shell.cleanDirs [ buildDir; nugetDir ])
 
+Target.create "ReplaceFsLibLogNamespaces"
```
Comment (TheAngryByrd) [inline build.fsx:93]: Added FsLibLog to help debugging in FSAC.

File: src/Ionide.ProjInfo.ProjectSystem/ProjectSystem.fs [hunk 2]
```diff
@@ -36,7 +36,7 @@ type ProjectResponse =
 /// Public API for any operations related to workspace and projects.
 /// Internally keeps all the information related to project files in current workspace.
 /// It's responsible for refreshing and caching - should be used only as source of information and public API
-type ProjectController(toolsPath: ToolsPath) as x =
+type ProjectController(toolsPath: ToolsPath, workspaceLoaderFactory: ToolsPath -> IWorkspaceLoader) as x =
```
Comment (TheAngryByrd) [inline src/Ionide.ProjInfo.ProjectSystem/ProjectSystem.fs:39]: This now takes a function to create a workspace loader.  Allowing us to keep the current Loader but letting people trial the new one.

File: src/Ionide.ProjInfo/Library.fs [hunk 3]
```diff
@@ -30,7 +31,7 @@ module Init =
 /// In most cases you want to use `Ionide.ProjInf.WorkspaceLoader` type instead
 module ProjectLoader =
 
-    type LoadedProject = private LoadedProject of ProjectInstance
+    type LoadedProject = internal LoadedProject of ProjectInstance
```
Comment (TheAngryByrd) [inline src/Ionide.ProjInfo/Library.fs:34]: I made this internal so `WorkspaceLoaderViaProjectGraph` but I could move the logic into this module so this won't get exposed again.

File: src/Ionide.ProjInfo/Library.fs [hunk 4]
```diff
@@ -65,25 +66,44 @@ module ProjectLoader =
         else
             None
 
+    let createLoggers (paths: string seq) (generateBinlog: bool) (sw: StringWriter) =
```
Comment (TheAngryByrd) [inline src/Ionide.ProjInfo/Library.fs:69]: Extracted a bunch of similar shared logic

File: src/Ionide.ProjInfo/Library.fs [hunk 5]
```diff
@@ -291,120 +345,282 @@ module ProjectLoader =
...
+            pg
+
+
+
+
+
+
+    let loadProjects (projects: ProjectGraph, customProperties: string list, generateBinlog: bool) =
+        try
+            lock WorkspaceLoaderViaProjectGraph.locker
```
Comment (TheAngryByrd) [inline src/Ionide.ProjInfo/Library.fs:417]: Currently does a lock so we don't try to run more than one build (msbuild doesn't like that).  Could change this to a MailboxProcessor or Semaphor if desired.

File: src/Ionide.ProjInfo/Library.fs [hunk 6]
```diff
@@ -291,120 +345,282 @@ module ProjectLoader =
...
 
-                Result.Ok proj
-        | Error e -> Result.Error e
+open Ionide.ProjInfo.Logging
+
+module WorkspaceLoaderViaProjectGraph =
+    let locker = obj ()
+
+
+type IWorkspaceLoader =
```
Comment (TheAngryByrd) [inline src/Ionide.ProjInfo/Library.fs:364]: Shared interface between old and new loader.

File: test/Ionide.ProjInfo.Tests/Tests.fs [hunk 7]
```diff
@@ -106,23 +106,24 @@ module ExpectNotification =
...
         |> List.iter
-            (fun (n, check) ->
-                let name, f = check
-
-                let minimal_info =
-                    match n with
-                    | WorkspaceProjectState.Loading (path) -> sprintf "loading %s " path
-                    | WorkspaceProjectState.Loaded (po, _, _) -> sprintf "loaded %s" po.ProjectFileName
-                    | WorkspaceProjectState.Failed (path, _) -> sprintf "failed %s" path
+            (fun (name, f) ->
```
Comment (TheAngryByrd) [inline test/Ionide.ProjInfo.Tests/Tests.fs:120]: The new graph solution may order messages slightly different due to it's sorting. I had to make this more flexible with items potentially being out of order.

File: test/Ionide.ProjInfo.Tests/Tests.fs [hunk 8]
```diff
@@ -955,30 +948,77 @@ let debugTets toolsPath =
...
         "Main tests"
-        [ testSample2 toolsPath
-          testSample3 toolsPath //- Sample 3 having issues, was also marked pending on old test suite
-          testSample4 toolsPath
-          testSample5 toolsPath
-          testSample9 toolsPath
+        [ testSample2 toolsPath "WorkspaceLoader" WorkspaceLoader.Create
+          testSample2 toolsPath "WorkspaceLoaderViaProjectGraph" WorkspaceLoaderViaProjectGraph.Create
+          testSample3 toolsPath "WorkspaceLoader" WorkspaceLoader.Create testSample3WorkspaceLoaderExpected //- Sample 3 having issues, was also marked pending on old test suite
+          //   testSample3 toolsPath "WorkspaceLoaderViaProjectGraph" WorkspaceLoaderViaProjectGraph.Create testSample3GraphExpected //- Sample 3 having issues, was also marked pending on old test suite
```
Comment (TheAngryByrd) [inline test/Ionide.ProjInfo.Tests/Tests.fs:987]: testSample3 was giving me some problems.  I can work on this more if desired.

File: test/Ionide.ProjInfo.Tests/Tests.fs [hunk 9]
```diff
@@ -955,30 +948,77 @@ let debugTets toolsPath =
...
+    let testSlnExpected =
+        [ ExpectNotification.loading "c1.fsproj"
+          ExpectNotification.loading "l2.fsproj"
+          ExpectNotification.loaded "l2.fsproj"
+          ExpectNotification.loaded "c1.fsproj"
+          ExpectNotification.loading "l1.fsproj"
+          ExpectNotification.loaded "l1.fsproj"
+          ExpectNotification.loaded "l2.fsproj" ]
+
+    let testSlnGraphExpected =
```
Comment (TheAngryByrd) [inline test/Ionide.ProjInfo.Tests/Tests.fs:972]: We won't get duplicate messages anymore so I had to pivot on expectations for either implementation

**Review by TheAngryByrd — COMMENTED** [review]
Self-review pointing out significant changes

**Comment by TheAngryByrd** [comment]
I think this is ready to go, just needs review/approval.

File: src/Ionide.ProjInfo.ProjectSystem/ProjectSystem.fs (hunk 2 above)
Comment (baronfel) [reply src/Ionide.ProjInfo.ProjectSystem/ProjectSystem.fs:39]: I really like this 👍

**Review by baronfel — APPROVED** [review]
This looks great to me, thank you for this work!
